clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -65,17 +66,91 @@
  */
 public class TransactionManager {
     private static final int NO_INFLIGHT_REQUEST_CORRELATION_ID = -1;
+    private static final int NO_LAST_ACKED_SEQUENCE_NUMBER = -1;
 
     private final Logger log;
     private final String transactionalId;
     private final int transactionTimeoutMs;
 
-    // The base sequence of the next batch bound for a given partition.
-    private final Map<TopicPartition, Integer> nextSequence;
+    private static class TopicPartitionBookeeper {
```

Comment:
nit: `class TopicPartitionBookKeeper` ?

Reply:
I think we misunderstood it. :) ![bookeeper](https://user-images.githubusercontent.com/1820518/52848737-c02dac80-310f-11e9-8f74-15b4552a167b.png)
(Thanks though, I'll fix the typo. And sorry for the bad joke :) )